### PR TITLE
feat: Added trainer for binary classification

### PR DESCRIPTION
**Pull request by MateoLostanlen**
Creation of a Binary Classification Trainer according to the discussion #108

### CODE REVIEW

File: holocron/trainer/core.py [hunk 1]
```diff
@@ -364,6 +364,52 @@ def _eval_metrics_str(eval_metrics: Dict[str, float]) -> str:
...
+        val_loss, top1, num_samples = 0., 0, 0
+        for x, target in self.val_loader:
+            x, target = self.to_cuda(x, target)
+
+            # Forward
+            out = self.model(x)
+            # Loss computation
+            val_loss += self.criterion(out, target).item()
+
+            top1 += torch.sum(abs(target - torch.sigmoid(out)) < 0.5).item()
```
Comment (frgfm) [inline holocron/trainer/core.py:399]: Mmmh, I see two things I'm not sure how to handle:
- in the other trainers, you don't assume how the predictions are made (whether it's a sigmoid or something else). I'm not sure how we could achieve it here (perhaps a use_sigmoid flag set in the constructor?)
- so the `abs(target - torch.sigmoid(out)) < 0.5` is actually dangerous, because this assumes that the target is binary. While the source target  is, if the user add something like label smoothing you can end up with a target of 0.9. And with your criteria, if the predicted probability was 0.41, you would wrongly say that the prediction was correct. I would suggest `(target >=0.5) == (torch.sigmoid(out) >= 0.5)` which would solve the issue.

File: test/test_trainer.py [hunk 2]
```diff
@@ -132,6 +132,24 @@ def test_classification_trainer(self):
...
+        batch_size = 8
+        # Generate all dependencies
+        model = nn.Sequential(nn.Conv2d(3, 32, 3), nn.ReLU(inplace=True),
+                              GlobalAvgPool2d(flatten=True), nn.Linear(32, 1))
+        train_loader = torch.utils.data.DataLoader(MockClassificationDataset(num_it * batch_size),
+                                                   batch_size=batch_size)
+        optimizer = torch.optim.Adam(model.parameters())
+        criterion = torch.nn.CrossEntropyLoss()
+
+        learner = trainer.BinaryClassificationTrainer(model, train_loader, train_loader, criterion, optimizer)
```
Comment (frgfm) [inline test/test_trainer.py:147]: There is an issue here: if this is for binary classification, it needs a proper loss (BCE rather than Cross entropy), and a proper dataset (MockClassificationDataset is for multi-class classification).

File: test/test_trainer.py [hunk 3]
```diff
@@ -132,6 +132,24 @@ def test_classification_trainer(self):
...
+
+        res = learner.evaluate()
+        # Random test here, juste checking if evaluate ended well
+        self.assertGreater(res['acc'], -1)
```
Comment (frgfm) [inline test/test_trainer.py:151]: No need to make this random: the accuracy is expected to be a float between 0 and 1, we can easily make this test :)

**Review by frgfm — CHANGES_REQUESTED** [review]
Thanks for the PR! I added some comments, let me know what you think!

File: holocron/trainer/core.py (hunk 1 above)
Comment (MateoLostanlen) [reply holocron/trainer/core.py:399]: In both cases I was a little too focused on my use case. But yes the flag in the constructor seems to me insdispensable and I will add the modification for the second point. Thanks !

File: test/test_trainer.py (hunk 2 above)
Comment (MateoLostanlen) [reply test/test_trainer.py:147]: I don't get your point here, MockClassificationDataset returns an image and a target to 0. 0 can either be a class identifier or a probability.

File: holocron/trainer/core.py (hunk 1 above)
Comment (MateoLostanlen) [reply holocron/trainer/core.py:399]: ok I get it know, is about the size of the output tensor :/

File: test/test_trainer.py (hunk 2 above)
Comment (MateoLostanlen) [reply test/test_trainer.py]: ok I get it know, is about the size of the output tensor :/

**Comment by frgfm** [comment]
You need to merge master, now that your other PR has been merged @MateoLostanlen ;)

File: holocron/trainer/core.py [hunk 4]
```diff
@@ -366,6 +366,62 @@ def _eval_metrics_str(eval_metrics: Dict[str, float]) -> str:
...
+        train_loader (torch.utils.data.DataLoader): training loader
+        val_loader (torch.utils.data.DataLoader): validation loader
+        criterion (torch.nn.Module): loss criterion
+        optimizer (torch.optim.Optimizer): parameter optimizer
+        gpu (int, optional): index of the GPU to use
+        output_file (str, optional): path where checkpoints will be saved
+        use_sigmoid (bool, optional): apply sigmoid function to the output
+    """
+
+    def __init__(self, model, train_loader, val_loader, criterion, optimizer, gpu=None, output_file='./checkpoint.pth',
```
Comment (frgfm) [inline holocron/trainer/core.py:383]: If we manage to find a way to avoid modifying the constructor, that would be even better. Perhaps setting the attribute by default (not in the constructor)?

If not, we need to add some proper typing here

File: test/test_trainer.py [hunk 5]
```diff
@@ -155,6 +168,24 @@ def test_classification_trainer_evaluate(self):
...
+
+        num_it = 10
+        batch_size = 8
+        # Generate all dependencies
+        model = nn.Sequential(nn.Conv2d(3, 32, 3), nn.ReLU(inplace=True),
+                              GlobalAvgPool2d(flatten=True), nn.Linear(32, 1))
+        train_loader = torch.utils.data.DataLoader(MockBinaryClassificationDataset(num_it * batch_size),
+                                                   batch_size=batch_size)
+        optimizer = torch.optim.Adam(model.parameters())
+        criterion = torch.nn.BCELoss()
```
Comment (frgfm) [inline test/test_trainer.py:181]: We might have to make a proper test, because this wouldn't converge properly : the BCELoss assumes a prediction in the range [0, 1] during training. Considering the architecture in use, this isn't the case

I'd recommend switching to BCEWithLogitsLoss https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html

File: holocron/trainer/core.py (hunk 4 above)
Comment (frgfm) [reply holocron/trainer/core.py:383]: actually for now, this would be simpler to use the torch.sigmoid by default in the eval run. I'll think about how to handle this non-sigmoid option later on

**Review by frgfm — CHANGES_REQUESTED** [review]
Sorry I actually changed my mind, adding the flag in the constructor sounds like a big change of interface for now. Let's put the sigmoid back by default, we'll change that later.

**Comment by MateoLostanlen** [comment]
> 
> 
> Sorry I actually changed my mind, adding the flag in the constructor sounds like a big change of interface for now. Let's put the sigmoid back by default, we'll change that later.

If you are not comfortable with this PR, I can add this BinaryTrainer to Pyro-vision instead

**Review by frgfm — APPROVED** [review]
Thanks for the PR!

**Comment by frgfm** [comment]
@MateoLostanlen no don't worry, it's useful! I'll have to think about how to make it more flexible later

I approved, but if you could also add the BinaryClassificationTrainer to the documentation, that would be great: https://github.com/frgfm/Holocron/blob/master/docs/source/trainer.rst

**Comment by MateoLostanlen** [comment]
> 
> 
> @MateoLostanlen no don't worry, it's useful! I'll have to think about how to make it more flexible later
> 
> I approved, but if you could also add the BinaryClassificationTrainer to the documentation before merging, that would be great: https://github.com/frgfm/Holocron/blob/master/docs/source/trainer.rst

done :)

**Review by frgfm — APPROVED** [review]
Thanks for the edit!
